### PR TITLE
Use lock-free algorithms in OptionServiceFactory.OptionService

### PR DESCRIPTION
**Pull request by sharwell**
<details><summary>Ask Mode not yet complete</summary><p>

**Customer scenario**

What does the customer do to get into this situation, and why do we think this
is common enough to address for this release.  (Granted, sometimes this will be
obvious "Open project, VS crashes" but in general, I need to understand how
common a scenario is)

**Bugs this fixes:**

(either VSO or GitHub links)

**Workarounds, if any**

Also, why we think they are insufficient for RC vs. RC2, RC3, or RTW

**Risk**

This is generally a measure our how central the affected code is to adjacent
scenarios and thus how likely your fix is to destabilize a broader area of code

**Performance impact**

(with a brief justification for that assessment (e.g. "Low perf impact because no extra allocations/no complexity changes" vs. "Low")

**Is this a regression from a previous update?**

**Root cause analysis:**

How did we miss it?  What tests are we adding to guard against it in the future?

**How was the bug found?**

(E.g. customer reported it vs. ad hoc testing)

</p></details>

### CODE REVIEW

File: src/Workspaces/Core/Portable/Options/OptionServiceFactory.cs [hunk 1]
```diff
@@ -125,20 +117,28 @@ public OptionSet GetOptions()
...
             public void RegisterDocumentOptionsProvider(IDocumentOptionsProvider documentOptionsProvider)
             {
-                lock (_gate)
+                InterlockedUpdate(
+                    ref _documentOptionsProviders,
+                    (original, added) => original.Add(added),
+                    documentOptionsProvider);
+            }
+
+            private static void InterlockedUpdate<T, TArg>(ref ImmutableArray<T> array, Func<ImmutableArray<T>, TArg, ImmutableArray<T>> update, TArg argument)
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Options/OptionServiceFactory.cs]: ❓ I was surprised to find this not already a part of `ImmutableInterlocked`. Do we have a good place to put this method in the current code base?

Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Options/OptionServiceFactory.cs]: The compiler does a lot of stuff like this, they might have a shared place.

**Comment by jasonmalinowski** [comment]
What's the measured cost of this lock as it is today? Just curious.

**Comment by sharwell** [comment]
@jasonmalinowski This one wasn't showing up on the profiler traces, because the calls are made from an `AsyncLazy` which is itself guarded.

File: src/Workspaces/Core/Portable/Options/OptionServiceFactory.cs [hunk 2]
```diff
@@ -40,11 +40,6 @@ internal class OptionService : IWorkspaceOptionService
             private readonly IGlobalOptionService _globalOptionService;
             private readonly IWorkspaceTaskScheduler _taskQueue;
 
-            /// <summary>
-            /// Gate guarding <see cref="_eventHandlers"/> and <see cref="_documentOptionsProviders"/>.
-            /// </summary>
-            private readonly object _gate = new object();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Options/OptionServiceFactory.cs]: is this collection being mutated so much that this gate was a problem?  I'm curious where the contention is coming from.

**Comment by jasonmalinowski** [comment]
@sharwell: so then why make the change?

**Comment by sharwell** [comment]
@jasonmalinowski It was an unnecessary usage of `lock` that I found on the fast path.

**Comment by jasonmalinowski** [comment]
I'm trying to reconcile these two statements:

> This one wasn't showing up on the profiler traces

> the less of them I'll be seeing in PerfView

Was this one that we did see or not?

**Comment by sharwell** [comment]
@jasonmalinowski This one did not show up. I located the inefficiency through code review in a related area.

**Comment by jasonmalinowski** [comment]
If the only concern is the contention on the lock in `GetUpdatedOptionSetForDocumentAsync` when reading, let's just remove that one lock, which is OK since the types are immutable. (I'm wondering if that was my intent moving them to be immutable and just forgot to remove the one lock.)

I don't see the value in doing fancy lockfree stuff for some of those methods that are called once or twice per run of Visual Studio. (For me lock-free code starts at minus 1000 points in a lot of cases...) Let's just keep that code simple.

**Comment by sharwell** [comment]
@jasonmalinowski It's not particularly fancy in this case. [`ImmutableInterlocked.Update`](https://msdn.microsoft.com/en-us/library/system.collections.immutable.immutableinterlocked.update.aspx) is the standard way for updating immutable collections in a concurrent context. It just appears they left out the overload for `ImmutableArray<T>`.

**Comment by jasonmalinowski** [comment]
Anything lock free is fancy in my book. :smile: I'd prefer the dumb lock here, but just do the read outside the lock with appropriate commenting. It's a pattern we do in lots of other places.

**Comment by sharwell** [comment]
> It's a pattern we do in lots of other places.

I am very aware of this, thanks to PerfView. 😑

**Comment by sharwell** [comment]
@jaredpar Will reopen if problems are observed

**Comment by jaredpar** [comment]
Thanks!

**Pull request closed without merging**